src/libCli/Completion.cpp

```diff
@@ -19,198 +19,251 @@ using namespace ArgParse;
 
 namespace cli
 {
-
-void printFishCompletions( std::vector<std::shared_ptr<ParsedElement> > & f_candidates, ParsedElement & f_parseTree, const std::string & f_args, bool f_debug)
-{
-    // completion requested :)
-    if(f_debug)
+    std::string getLastArgument(std::string &candStr, const char *delim, std::vector<std::string> &sugs)
     {
-        std::cerr << "Input string \"" << f_args << std::endl;
-    }
 
-    //size_t n = parseTree.getMatchedString().size();
-    size_t n = f_args.size();
-    for(auto candidate : f_candidates)
-    {
-        std::string candidateStr = candidate->getMatchedStringRaw();
+        //char *candArg = const_cast<char *>(ca,ndStr.c_str());
+        std::vector<std::string> candSuggestion;
 
-        if(f_debug)
-        {
-            std::cout << "------------------------------------------------" << std::endl;
-            std::cout << "candiate string: " << candidateStr << std::endl;
-            std::cout << "candiate's debug string: " << std::endl;
-            std::cout << candidate->getDebugString() << std::endl;
-            std::cout << "------------------------------------------------" << std::endl;
-        }
+        //if (sugs.back() == ":")
+        //{
+        //return sugs[sugs.size() - 1];
+        //candSuggestion.push_back();
+        //    return "*****Shalalala***********";
+        //}
 
-        std::string suggestion;
-        size_t start = n;
-        size_t end;
+        // split Candidate at given delimiter
+        char *candArg = strtok(const_cast<char *>(candStr.c_str()), delim);
```

Comment:
from this point on, behavior of the program is undefined.
never use const_cast.

What is happening here:
`candStr` is of type std::string
strtok expects an argument of type `char*` which is the "old" C style of representing strings.

Those two types have very little in common.
However C++ std::String provides some kind of backwards compatibility: `.c_str()` which returns a `const char *` which looks to a C program like an old-style C String.
However note the `const`.
This means the code using the returned `char*` is allowed to read but not write. The std::string specification says that any modification to the data pointed to by the returned pointer leads to undefined behavior.

The C++ compiler enforces this and disallows writes (implicitly via function signatures). This is a nice thing and prevents errors.

strtok will try to write to the C-style string passed as first argument. Hence this is not permitted and compiler will not compile.

by using `const_cast` here you are overriding the compiler error and just continue. Most people are not better than compiler in judging if something is OK or not. So never use const_cast :P

---

you have two choices to fix this:
1. stick to std::string types and corresponding functions only:
   https://en.cppreference.com/w/cpp/string/basic_string
   or the std::regex lib if string operations get more complex.
2. If you really need to use modifying functions which accept old-style strings and cannot find an alternative, you will at some point need to copy the string to heap memory (maybe via vector) in a very controlled manner. Dangerous way: `malloc` or `new`. Slightly better way: `std::vector` with `resize()` member function.